hw/xbox/mcpx_apu.c

```diff
@@ -473,6 +514,120 @@ static void ep_scratch_rw(void *opaque,
                       ptr, addr, len, dir);
 }
 
+static uint32_t circular_scatter_gather_rw(MCPXAPUState *d,
+                                           hwaddr sge_base,
+                                           unsigned int max_sge,
+                                           uint8_t *ptr,
+                                           uint32_t base, uint32_t end,
```

Comment:
Is `end` supposed to be inclusive or exclusive?

Reply:
I don't remember but the boundaries should be correct, as this implementation was tested on hardware.

Anyhow, this code strongly implies it's exclusive (`end - cur` or `cur >= end`).

As these are hardware registers (`NV_PAPU_xPxFENDx`), we should document the inclusive / exclusive stuff on XboxDevWiki.
I'm not sure if documenting in the code is necessary (given the strong implications + better docs in the wiki). We also don't document similar stuff in nv2a if it's well defined in other docs (OpenGL specs for example).

Reply:
The address is exclusive.

Confirmed on hardware, and also makes a lot of sense as the lower bits of these registers are not stored (which always forces a 0x100 aligned address).